crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -1233,59 +1248,105 @@ impl Loader<HistoricalKey> for Db {
+                    }
+
+                    query
+                })
+                .await
+            }
+            .scope_boxed()
+        });
+        let (indexed_objects_result, native_objects_result) =
+            futures::join!(indexed_objects_query, native_object_query);
+        let indexed_objects: Vec<StoredHistoryObject> = indexed_objects_result
+            .map_err(|e| Error::Internal(format!("Failed to fetch objects: {e}")))?;
+        let native_objects: Vec<StoredFullHistoryObject> = native_objects_result
             .map_err(|e| Error::Internal(format!("Failed to fetch objects: {e}")))?;
 
-        let mut id_version_to_stored = BTreeMap::new();
-        for stored in objects {
-            let key = (addr(&stored.object_id)?, stored.object_version as u64);
-            id_version_to_stored.insert(key, stored);
+        let mut id_version_to_indexed = BTreeMap::new();
+        for indexed in indexed_objects {
+            let key = (addr(&indexed.object_id)?, indexed.object_version as u64);
+            id_version_to_indexed.insert(key, indexed);
+        }
+        let mut id_version_to_native = BTreeMap::new();
+        for native in native_objects {
+            let key = (addr(&native.object_id)?, native.object_version as u64);
+            id_version_to_native.insert(key, native);
         }
 
         let mut result = HashMap::new();
         for key in keys {
-            let Some(stored) = id_version_to_stored.get(&(key.id, key.version)) else {
-                continue;
-            };
+            let indexed_opt = id_version_to_indexed.get(&(key.id, key.version));
 
             // Filter by key's checkpoint viewed at here. Doing this in memory because it should be
             // quite rare that this query actually filters something, but encoding it in SQL is
             // complicated.
-            if key.checkpoint_viewed_at < stored.checkpoint_sequence_number as u64 {
-                continue;
-            }
+            let object = if let Some(indexed) = indexed_opt {
+                if key.checkpoint_viewed_at < indexed.checkpoint_sequence_number as u64 {
+                    continue;
+                }
+                Object::try_from_stored_history_object(
+                    indexed.clone(),
+                    key.checkpoint_viewed_at,
+                    // This conversion will use the object's own version as the `Object::root_version`.
+                    None,
+                )?
+            } else {
+                let Some(native) = id_version_to_native.get(&(key.id, key.version)) else {
+                    continue;
+                };
+                if let Some(serialized) = &native.serialized_object {
+                    let object = bcs::from_bytes(serialized).map_err(|_| {
+                        Error::Internal(format!("Failed to deserialize object at {:?}", key))
+                    })?;
+                    Object::from_native(
```

Comment:
`from_native` is probably not the right constructor to use here, because:

- it will result in the object status being "not indexed",
- it introduces some avoidable inefficiency in that if you request the BCS of the object, it will reserialize it rather than offer the BCS that we were storing in the first place.

For now, the most straightforward thing would be to introduce yet another `ObjectKind` for this kind of point lookup, and we can revisit those kinds in general at a later date. I think in the end they will look quite different, and that needs to be carefully designed.